File: src/main/java/com/customagent/service/AgentService.java
package com.customagent.service;

import com.customagent.exception.AgentNotFoundException;
import com.customagent.model.Agent;
import com.customagent.model.AgentStatus;
import com.customagent.model.dto.AgentRequest;
import com.customagent.repository.AgentRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class AgentService {

    private final AgentRepository agentRepository;

    // ─── CREATE ──────────────────────────────────────────────────────────────

    /**
     * Creates a new agent.
     *
     * @throws IllegalArgumentException if the name is already taken
     */
    public Agent createAgent(AgentRequest request) {
        if (agentRepository.existsByName(request.getName())) {
            throw new IllegalArgumentException(
                    "Agent with name '" + request.getName() + "' already exists");
        }

        Agent agent = Agent.builder()
                .name(request.getName())
                .description(request.getDescription())
                .tools(request.getTools() != null
                        ? request.getTools()
                        : Collections.emptyList())
                .status(AgentStatus.ACTIVE)
                .build();

        Agent saved = agentRepository.save(agent);
        log.info("Created agent id={} name={}", saved.getId(), saved.getName());
        return saved;
    }

    // ─── READ ─────────────────────────────────────────────────────────────────

    @Transactional(readOnly = true)
    public Agent getAgentById(Long id) {
        return agentRepository.findById(id)
                .orElseThrow(() -> new AgentNotFoundException(
                        "Agent not found with id: " + id));
    }

    @Transactional(readOnly = true)
    public List<Agent> getAllAgents() {
        return agentRepository.findAll();
    }

    // ─── UPDATE ───────────────────────────────────────────────────────────────

    /**
     * Fully replaces the mutable fields of an existing agent.
     *
     * @throws AgentNotFoundException   if {@code id} does not exist
     * @throws IllegalArgumentException if the new name is already taken by another agent
     */
    public Agent updateAgent(Long id, AgentRequest request) {
        Agent existing = getAgentById(id);

        boolean nameChanged = !existing.getName().equals(request.getName());
        if (nameChanged && agentRepository.existsByName(request.getName())) {
            throw new IllegalArgumentException(
                    "Agent with name '" + request.getName() + "' already exists");
        }

        existing.setName(request.getName());
        existing.setDescription(request.getDescription());
        existing.setTools(request.getTools() != null
                ? request.getTools()
                : Collections.emptyList());

        Agent updated = agentRepository.save(existing);
        log.info("Updated agent id={}", updated.getId());
        return updated;
    }

    // ─── DELETE ───────────────────────────────────────────────────────────────

    public void deleteAgent(Long id) {
        Agent agent = getAgentById(id);
        agentRepository.delete(agent);
        log.info("Deleted agent id={}", id);
    }

    // ─── STATUS TRANSITIONS ───────────────────────────────────────────────────

    public Agent activateAgent(Long id) {
        Agent agent = getAgentById(id);
        agent.setStatus(AgentStatus.ACTIVE);
        return agentRepository.save(agent);
    }

    public Agent deactivateAgent(Long id) {
        Agent agent = getAgentById(id);
        agent.setStatus(AgentStatus.INACTIVE);
        return agentRepository.save(agent);
    }
}

File: src/services/agentService.js
'use strict';

/**
 * In-memory Agent store with CRUD + lifecycle operations.
 * Mirrors the Java AgentService business logic.
 *
 * In production this would delegate to a database; here an injectable
 * `repository` abstraction makes unit-testing straightforward.
 */

const AgentStatus = Object.freeze({ ACTIVE: 'ACTIVE', INACTIVE: 'INACTIVE' });

let _nextId = 1;

/**
 * Creates an AgentService bound to the provided repository.
 *
 * @param {{ findById, findAll, existsByName, save, delete: deleteById }} repository
 * @returns {object} service instance
 */
function createAgentService(repository) {
  // ── CREATE ──────────────────────────────────────────────────────────────

  /**
   * Creates a new agent.
   *
   * @param {{ name: string, description?: string, tools?: string[] }} request
   * @returns {object} Saved agent
   * @throws {Error} If name is already taken
   */
  function createAgent(request) {
    if (!request || !request.name || request.name.trim() === '') {
      throw new Error('Agent name is required');
    }
    if (request.name.trim().length < 2 || request.name.trim().length > 100) {
      throw new Error('Name must be between 2 and 100 characters');
    }
    if (repository.existsByName(request.name)) {
      throw new Error(`Agent with name '${request.name}' already exists`);
    }

    const agent = {
      id: _nextId++,
      name: request.name.trim(),
      description: request.description || null,
      tools: Array.isArray(request.tools) ? [...request.tools] : [],
      status: AgentStatus.ACTIVE,
      createdAt: new Date().toISOString(),
    };

    return repository.save(agent);
  }

  // ── READ ─────────────────────────────────────────────────────────────────

  /**
   * @param {number} id
   * @returns {object} Agent
   * @throws {Error} If not found
   */
  function getAgentById(id) {
    const agent = repository.findById(id);
    if (!agent) {
      throw new Error(`Agent not found with id: ${id}`);
    }
    return agent;
  }

  /**
   * @returns {object[]} All agents
   */
  function getAllAgents() {
    return repository.findAll();
  }

  // ── UPDATE ───────────────────────────────────────────────────────────────

  /**
   * Fully replaces mutable fields of an existing agent.
   *
   * @param {number} id
   * @param {{ name: string, description?: string, tools?: string[] }} request
   * @returns {object} Updated agent
   * @throws {Error} If not found or name conflict
   */
  function updateAgent(id, request) {
    const existing = getAgentById(id);

    if (!request || !request.name || request.name.trim() === '') {
      throw new Error('Agent name is required');
    }

    const nameChanged = existing.name !== request.name.trim();
    if (nameChanged && repository.existsByName(request.name)) {
      throw new Error(`Agent with name '${request.name}' already exists`);
    }

    existing.name = request.name.trim();
    existing.description = request.description || null;
    existing.tools = Array.isArray(request.tools) ? [...request.tools] : [];

    return repository.save(existing);
  }

  // ── DELETE ───────────────────────────────────────────────────────────────

  /**
   * @param {number} id
   * @throws {Error} If not found
   */
  function deleteAgent(id) {
    const agent = getAgentById(id);
    repository.delete(agent.id);
  }

  // ── STATUS TRANSITIONS ───────────────────────────────────────────────────

  /**
   * @param {number} id
   * @returns {object} Updated agent
   */
  function activateAgent(id) {
    const agent = getAgentById(id);
    agent.status = AgentStatus.ACTIVE;
    return repository.save(agent);
  }

  /**
   * @param {number} id
   * @returns {object} Updated agent
   */
  function deactivateAgent(id) {
    const agent = getAgentById(id);
    agent.status = AgentStatus.INACTIVE;
    return repository.save(agent);
  }

  return {
    createAgent,
    getAgentById,
    getAllAgents,
    updateAgent,
    deleteAgent,
    activateAgent,
    deactivateAgent,
  };
}

/** Reset auto-increment id (useful for testing). */
function resetIdCounter(start = 1) {
  _nextId = start;
}

module.exports = { createAgentService, AgentStatus, resetIdCounter };
